### ISSUE TITLE
Separate parsing from style guide generation

### ISSUE THREAD

**Issue by JohnAlbin** [kss-node/kss-node#365]
Right now the `kss` cli will parse the source files and generate a style guide all in one go.

We should make it easier for users to modify the `KssStyleGuide` object or save it as a JSON file before generating the style guide.

**Comment by JohnAlbin** [comment]
Part of this is done. https://github.com/kss-node/kss-node/issues/284 adds a `--json` option that will output the parsed style guide as a JSON string to standard output (if using the cli) or as a JSON object (if using kss.js).

**Comment by JohnAlbin** [comment]
Another part is now done in #374. You can now specify a .json file in `--source`.

**Comment by JohnAlbin** [comment]
Technically, you can use multiple commands to modify the KssStyleGuide object after parsing and before style guide generation. e.g.:

```
kss --source folder-of-css --json | unixUtil > kssObject.json
kss --source folder-of-css --source kssObject.json --destination styleguide
```

@ryuran Do you have any opinions on how we should allow modification of the KssStyleGuide object with one command?

**Comment by ryuran** [comment]
@JohnAlbin 
I think we should consider json as a default output of kss cli if no builder is installed. 
And the builder should transform this output to html css.

This allows kss users to build their own builder/theme.
This separate Kss, the API, and builder, the front builder.
Kss and builder could be streamed by gulp by exemple

**Comment by JohnAlbin** [comment]
> consider json as a default output of kss cli if no builder is installed.

Kss comes with 2 builders by default because I don't think people would know how to use the software without at least one. And more to your point, JSON output by default only makes sense for advanced users, IMO. Since most people will not have their own builder when they first use kss-node. I feel having an optional `--json` flag that turns on JSON output is better UX.

> This allows kss users to build their own builder/theme.

This is already true.

1. You can build one from scratch and use the `--builder [path]` to point at your custom builder. or
2. You can use the `--clone --builder [path]` option to clone an existing builder and hack away to create your own.
3. You can use the JavaScript API to output a `KssStyleGuide` object which you can use as-is or you can use `KssStyleGuide.toJSON` method to convert it into a JSON object. This would allow your custom Node.js code to call `kss()` with the `json` option and receive a `Promise` to the KssStyleGuide object that was created when parsing your source directories.

The docs definitely need improving in this area.

> Kss and builder could be streamed with gulp for example.

I absolutely want to do that. I'd prefer kss-node to have a built-in simple way to integrate with Gulp, but I've been waiting (for over a year) on Gulp 4 to stabilize the APIs that kss-node would need to use. See #226 

> This separate Kss, the API, and builder, the front builder.

I think we've achieve this already. And I absolutely am up for improvements in the separation. The `kss` command line utility is just a tiny wrapper around the `cli.js` code. And the `cli.js` code just bundles all the command line flags, handles a couple cli-specific options and then passes all the options to `kss()` so that it can either `traverse()` directories or read a JSON file and then either output a KssStyleGuide object or build a style guide. This also allows a custom parsing tool that generates its own KssStyleGuide or JSON object which can be fed to `kss()` to do the building.

**Comment by ryuran** [comment]
> Kss comes with 2 builders by default because I don't think people would know how to use the software without at least one. And more to your point, JSON output by default only makes sense for advanced users, IMO. Since most people will not have their own builder when they first use kss-node. I feel having an optional --json flag that turns on JSON output is better UX.

My point is kss have currently lot of dependencies (handlebars and twig.js for example) when it is not there own dependencies but builder's dependencies.

A twig user does not need handlebars.
A nunjucks user does not need twig.js.
If in the installation doc you can read how to install default builder `npm install kss-buider-twig` and `"builder": "kss-builder-twig"` in kss.json

You may throw an alert in cli if there are no builder.

> 1 - You can build one from scratch and use the --builder [path] to point at your custom builder. or
> 2 - You can use the --clone --builder [path] option to clone an existing builder and hack away to create your own.

But you can't fork original builder to keep your own up to date with the original.

**Comment by ryuran** [comment]
Or maybe kss parser (comments to API) can be a dependency of kss builder (API to html/css).

**Comment by JohnAlbin** [comment]
@ryuran Thanks for the discussion points. I think I have a solution that gets at what you want.

What if we split the current `kss` package into separate parts:

1. `kss-core` package: Contains `cli`, `kss`, traverse/parse utilities, and `KssBuilderBase`. By itself, it can only output JSON to stdout or, when using `kss()`, a `Promise` to a `KssStyleGuide` object.
2. `kss` package: Contains very little code, but has dependencies on `kss-core`, `kss-handlebars`, and `kss-twig`. When this package is installed, running the `kss` cli tool will use the `kss-handlebars` builder by default and optionally have the ability to use the twig builder.
3. `kss-handlebars` package: Contains `KssBuilderBaseHandlebars` and the Handlebars builder. Has a dependency on `kss-core`. When this package is installed, running the `kss` cli tool will use the Handlebars builder by default.
4. `kss-twig` package: Contains `KssBuilderBaseTwig` and the Twig builder. Has a dependency on `kss-core`. When this package is installed, running the `kss` cli tool will use the Twig builder by default.

Without actually looking at the code before typing this, I think the tricky part will be getting the `kss-twig` package to over-ride the default builder. But I think I can get that to work.

As a bonus, this will make it easier for maintainers of custom builders to instruct their users. If their custom builder package is set up the same way, they can say "install kss-custom and then kss will automatically use the custom builder when run."

What do you think?

**Comment by ryuran** [comment]
@JohnAlbin Thanks for your answers and for having listened my opinions.
It really seems a good idea! I'm completely agree.

**Comment by JohnAlbin** [comment]
> they can say "install kss-custom and then kss will automatically use the custom builder when run."

It's looking like this is very tricky. I'll keep trying though.

**Comment by ryuran** [comment]
Why not use [Lerna](https://github.com/lerna/lerna) to manage builders packages on a monorepo ?

**Comment by JohnAlbin** [comment]
> Why not use Lerna to manage builders packages on a monorepo ?

Absolutely, KSS should use it. When I first attempted this nearly 3 years ago, I didn't know about Lerna. But it is the perfect tool for this problem.

**Comment by jameelmoses** [comment]
@ryuran @JohnAlbin any updates/progress on this? I'd really like to see these broken out so that the builders aren't necessarily coupled with the API
